From d2319fd92d7b7b682557434d681fbc5e331e9628 Mon Sep 17 00:00:00 2001
From: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
Date: Thu, 14 Jul 2022 18:30:00 +0300
Subject: [PATCH 15/62] xen_pvdev: Do not assume Dom0 when changing directory
 permissions

Instead of forcing the owner to domid 0, read the directory permissions
and add xen_domid with the given permission to it.

Note that for other than Dom0 domain (non toolstack domain) the "driver_domain"
property should be set in domain config file for the toolstack to create
required directories in advance.

Signed-off-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
---
Fixes for:
xen be core: xen be core: xs_set_permissions device-model/2/backends/console: failed
xs_set_permissions device-model/2/backends/console: failed
xen be core: xen be core: xs_set_permissions device-model/2/backends/vkbd: failed
xs_set_permissions device-model/2/backends/vkbd: failed
xen be core: xen be core: xs_set_permissions device-model/2/backends/9pfs: failed
xs_set_permissions device-model/2/backends/9pfs: failed
---
---
 hw/xen/xen_pvdev.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/hw/xen/xen_pvdev.c b/hw/xen/xen_pvdev.c
index 8ab458922a..302ea75427 100644
--- a/hw/xen/xen_pvdev.c
+++ b/hw/xen/xen_pvdev.c
@@ -60,9 +60,9 @@ void xen_config_cleanup(void)
 
 int xenstore_mkdir(char *path, int p)
 {
-    struct xs_permissions perms[2] = {
+    unsigned int num;
+    struct xs_permissions *tmp, perms[2] = {
         {
-            .id    = 0, /* set owner: dom0 */
         }, {
             .id    = xen_domid,
             .perms = p,
@@ -75,6 +75,14 @@ int xenstore_mkdir(char *path, int p)
     }
     xenstore_cleanup_dir(g_strdup(path));
 
+    tmp = xs_get_permissions(xenstore, 0, path, &num);
+    if (tmp == NULL) {
+        xen_pv_printf(NULL, 0, "xs_get_permissions %s: failed\n", path);
+        return -1;
+    }
+    perms[0].id = tmp[0].id; /* retain the owner */
+    free(tmp);
+
     if (!xs_set_permissions(xenstore, 0, path, perms, 2)) {
         xen_pv_printf(NULL, 0, "xs_set_permissions %s: failed\n", path);
         return -1;
-- 
2.34.1

